harness/determined/exec/tensorboard.py

```diff
@@ -95,60 +96,80 @@ def check_tensorboard_responsive() -> bool:
+
         with det.util.forward_signals(tensorboard_process):
             try:
-                while True:
-                    ret_code = tensorboard_process.poll()
-                    if ret_code is not None:
-                        raise RuntimeError(f"Tensorboard process died, exit code({ret_code}).")
+                tb_unresponsive_stop_time = time.time() + TB_RESPONSE_WAIT_TIME
+                print(TRIGGER_WAITING_MSG, flush=True)
+
+                # Wait for the Tensorboard process to start responding before proceeding.
+                responsive = False
+                while not responsive:
+                    raise_if_dead(tensorboard_process)
+
+                    if time.time() > tb_unresponsive_stop_time:
+                        raise RuntimeError("Tensorboard wasn't responsive before the timeout.")
 
-                    # Check if we have reached a timeout without receiving metrics
-                    if num_fetched_files == 0:
-                        if time.time() > stop_time:
-                            raise RuntimeError("No new files were fetched before the timeout.")
-                        else:
-                            print(TRIGGER_WAITING_MSG, flush=True)
+                    time.sleep(TICK_INTERVAL)
+                    responsive = check_tensorboard_responsive()
 
-                    if not responsive:
-                        if time.time() > stop_time:
-                            raise RuntimeError("Tensorboard wasn't responsive before the timeout.")
-                        responsive = check_tensorboard_responsive()
+                # Continuously loop checking for new files
+                stop_time = time.time() + MAX_WAIT_TIME
+                while True:
+                    raise_if_dead(tensorboard_process)
 
-                    if responsive and not triggered and num_fetched_files > 0:
-                        print(TENSORBOARD_TRIGGER_READY_MSG, flush=True)
-                        triggered = True
+                    # Check if we have reached a timeout without downloading any files
+                    if tb_fetch_manager.num_fetched_files == 0 and time.time() > stop_time:
+                        raise RuntimeError("No new files were fetched before the timeout.")
 
-                    time.sleep(FETCH_INTERVAL)
-                    num_fetched_files += fetcher.fetch_new()
+                    time.sleep(TICK_INTERVAL)
+                    # TODO: Note that this call is blocking and serial. We won't check
+                    # the stop time until this completely finishes
+                    fetcher.fetch_new(new_file_callback=tb_fetch_manager.on_file_fetched)
 
             finally:
                 if tensorboard_process.poll() is None:
                     logger.info("Killing tensorboard process")
                     tensorboard_process.kill()
 
-            return tensorboard_process.wait()
+
+class TBFetchManager:
+    def __init__(self):
+        self._ready = False
```

Comment:
just one nit: it might be easier to read with `_ready` but isn't `self._ready == (self.num_fetched_files == 0)` ? It seems to be an invariant.

Reply:
true, that's leftover code from attempts to set it at other values of fetched files, I'll remove it